### PR TITLE
Add linting for CSS in HTML

### PR DESCRIPTION
**Pull request by jfcherng**
image

Just found this. Settings are copied from `LSP-css`.

### CODE REVIEW

**Comment by rchl** [comment]
Unrelated but you could also remove `"enabled": true` from `LSP-html.sublime-settings` since it's not really read from there (it's supposed to be handled by removing or disabling package instead).

**Comment by rchl** [comment]
I'm getting some issue on one HTML file:
```
lsp-html: internal/modules/cjs/loader.js:638
lsp-html: throw err;
:: --> lsp-html initialize(1): {'initializationOptions': {'provideFormatter': True, 'embeddedLanguages': {'html': True, 'css': True, 'javascript': True}}, 'processId': 52351, 'rootUri': 'file:///Users/me/folder1', 'capabilities': {'workspace': {'symbol': {'symbolKind': {'valueSet': [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26]}}, 'applyEdit': True, 'configuration': True, 'executeCommand': {}, 'workspaceFolders': True, 'didChangeConfiguration': {}}, 'textDocument': {'hover': {'contentFormat': ['markdown', 'plaintext']}, 'formatting': {}, 'definition': {'linkSupport': True}, 'rename': {}, 'synchronization': {'didSave': True, 'willSaveWaitUntil': True, 'willSave': True}, 'references': {}, 'documentSymbol': {'symbolKind': {'valueSet': [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26]}}, 'completion': {'completionItem': {'snippetSupport': True}, 'completionItemKind': {'valueSet': [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25]}}, 'publishDiagnostics': {'relatedInformation': True}, 'implementation': {'linkSupport': True}, 'colorProvider': {}, 'rangeFormatting': {}, 'signatureHelp': {'signatureInformation': {'parameterInformation': {'labelOffsetSupport': True}, 'documentationFormat': ['markdown', 'plaintext']}}, 'typeDefinition': {'linkSupport': True}, 'documentHighlight': {}, 'codeAction': {'codeActionLiteralSupport': {'codeActionKind': {'valueSet': []}}}, 'declaration': {'linkSupport': True}}}, 'workspaceFolders': [{'uri': 'file:///Users/me/folder1', 'name': 'mikrokoszalin'}, {'uri': 'file:///Users/me/folder2', 'name': 'docker-php'}], 'rootPath': '/Users/me/folder1'}
:: --> lsp-html shutdown(2): None
:: --> lsp-html shutdown(3): None
```

But it doesn't seem new to this change... I have to investigate this later.

**Comment by rchl** [comment]
BTW. Here is VSCode configuration. It has formatting disabled, it seems but otherwise we should probably align.

<details>
  <summary>workspace/configuration</summary>
  
```json
{
    "settings": {
        "html": {
            "customData": [],
            "format": {
                "enable": false,
                "wrapLineLength": 120,
                "unformatted": "wbr",
                "contentUnformatted": "pre,code,textarea",
                "indentInnerHtml": false,
                "preserveNewLines": true,
                "maxPreserveNewLines": null,
                "indentHandlebars": false,
                "endWithNewline": false,
                "extraLiners": "head, body, /html",
                "wrapAttributes": "auto"
            },
            "suggest": {
                "html5": true
            },
            "validate": {
                "scripts": true,
                "styles": true
            },
            "autoClosingTags": true,
            "mirrorCursorOnMatchingTag": false,
        },
        "css": {
            "customData": [],
            "completion": {
                "triggerPropertyValueCompletion": true,
                "completePropertyWithSemicolon": true
            },
            "validate": true,
            "lint": {
                "compatibleVendorPrefixes": "ignore",
                "vendorPrefix": "warning",
                "duplicateProperties": "ignore",
                "emptyRules": "warning",
                "importStatement": "ignore",
                "boxModel": "ignore",
                "universalSelector": "ignore",
                "zeroUnits": "ignore",
                "fontFaceProperties": "warning",
                "hexColorLength": "error",
                "argumentsInColorFunction": "error",
                "unknownProperties": "warning",
                "validProperties": [],
                "ieHack": "ignore",
                "unknownVendorSpecificProperties": "ignore",
                "propertyIgnoredDueToDisplay": "warning",
                "important": "ignore",
                "float": "ignore",
                "idSelector": "ignore",
                "unknownAtRules": "warning"
            },
            "trace": {
                "server": "verbose"
            }
        },
        "javascript": {
            "referencesCodeLens": {
                "enabled": false,
                "showOnAllFunctions": false
            },
            "suggest": {
                "completeFunctionCalls": false,
                "includeAutomaticOptionalChainCompletions": true,
                "names": true,
                "paths": true,
                "autoImports": true,
                "completeJSDocs": true,
                "enabled": true
            },
            "validate": {
                "enable": true
            },
            "format": {
                "enable": true,
                "insertSpaceAfterCommaDelimiter": true,
                "insertSpaceAfterConstructor": false,
                "insertSpaceAfterSemicolonInForStatements": true,
                "insertSpaceBeforeAndAfterBinaryOperators": true,
                "insertSpaceAfterKeywordsInControlFlowStatements": true,
                "insertSpaceAfterFunctionKeywordForAnonymousFunctions": false,
                "insertSpaceBeforeFunctionParenthesis": false,
                "insertSpaceAfterOpeningAndBeforeClosingNonemptyParenthesis": false,
                "insertSpaceAfterOpeningAndBeforeClosingNonemptyBrackets": false,
                "insertSpaceAfterOpeningAndBeforeClosingNonemptyBraces": true,
                "insertSpaceAfterOpeningAndBeforeClosingTemplateStringBraces": false,
                "insertSpaceAfterOpeningAndBeforeClosingJsxExpressionBraces": false,
                "placeOpenBraceOnNewLineForFunctions": false,
                "placeOpenBraceOnNewLineForControlBlocks": false,
                "semicolons": "ignore"
            },
            "implicitProjectConfig": {
                "checkJs": true,
                "experimentalDecorators": false
            },
            "suggestionActions": {
                "enabled": true
            },
            "preferences": {
                "quoteStyle": "single",
                "importModuleSpecifier": "auto",
                "renameShorthandProperties": true
            },
            "updateImportsOnFileMove": {
                "enabled": "prompt"
            },
            "autoClosingTags": true
        }
    }
}
```</details>

**Comment by jfcherng** [comment]
That's really weird why they have formatting disabled for all servers :thinking:

**Comment by rchl** [comment]
Please also check my updated comment above with `initializationOptions`

**Comment by rchl** [comment]
> That's really weird why they have formatting disabled for all servers 🤔

They might have other means of formatting maybe. I think they still have prettier enabled by default but not sure.

**Comment by jfcherng** [comment]
After aligning settings/initializationOptions with VSC, now I see

```
:: --> lsp-html initialize(1): {'workspaceFolders': None, 'processId': 12688, 'capabilities': {'workspace': {'workspaceFolders': True, 'symbol': {'symbolKind': {'valueSet': [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26]}}, 'configuration': True, 'didChangeConfiguration': {}, 'executeCommand': {}, 'applyEdit': True}, 'textDocument': {'colorProvider': {}, 'references': {}, 'rangeFormatting': {}, 'formatting': {}, 'completion': {'completionItemKind': {'valueSet': [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25]}, 'completionItem': {'snippetSupport': True}}, 'documentSymbol': {'symbolKind': {'valueSet': [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26]}}, 'definition': {'linkSupport': True}, 'implementation': {'linkSupport': True}, 'publishDiagnostics': {'relatedInformation': True}, 'signatureHelp': {'signatureInformation': {'parameterInformation': {'labelOffsetSupport': True}, 'documentationFormat': ['markdown', 'plaintext']}}, 'hover': {'contentFormat': ['markdown', 'plaintext']}, 'rename': {}, 'codeAction': {'codeActionLiteralSupport': {'codeActionKind': {'valueSet': []}}}, 'typeDefinition': {'linkSupport': True}, 'declaration': {'linkSupport': True}, 'synchronization': {'willSaveWaitUntil': True, 'didSave': True, 'willSave': True}, 'documentHighlight': {}}}, 'rootPath': None, 'initializationOptions': {'provideFormatter': False, 'dataPaths': [], 'embeddedLanguages': {'css': True, 'javascript': True}}, 'rootUri': None}
:: <<< lsp-html 1: {'capabilities': {'colorProvider': {}, 'selectionRangeProvider': True, 'documentSymbolProvider': True, 'documentRangeFormattingProvider': False, 'renameProvider': True, 'foldingRangeProvider': True, 'textDocumentSync': 2, 'documentLinkProvider': {'resolveProvider': False}, 'documentHighlightProvider': True, 'referencesProvider': True, 'definitionProvider': True, 'completionProvider': {'resolveProvider': True, 'triggerCharacters': ['.', ':', '<', '"', '=', '/']}, 'hoverProvider': True, 'signatureHelpProvider': {'triggerCharacters': ['(']}}}
::  -> lsp-html workspace/didChangeConfiguration: {'settings': {'css': {'validate': True, 'customData': [], 'completion': {'triggerPropertyValueCompletion': True, 'completePropertyWithSemicolon': True}, 'lint': {'unknownProperties': 'warning', 'importStatement': 'ignore', 'hexColorLength': 'error', 'float': 'ignore', 'emptyRules': 'warning', 'validProperties': [], 'argumentsInColorFunction': 'error', 'unknownAtRules': 'warning', 'idSelector': 'ignore', 'zeroUnits': 'ignore', 'compatibleVendorPrefixes': 'ignore', 'vendorPrefix': 'warning', 'duplicateProperties': 'ignore', 'unknownVendorSpecificProperties': 'ignore', 'propertyIgnoredDueToDisplay': 'warning', 'ieHack': 'ignore', 'universalSelector': 'ignore', 'boxModel': 'ignore', 'important': 'ignore', 'fontFaceProperties': 'warning'}}, 'html': {'customData': [], 'mirrorCursorOnMatchingTag': False, 'format': {'indentInnerHtml': False, 'maxPreserveNewLines': None, 'enable': False, 'contentUnformatted': 'pre,code,textarea', 'indentHandlebars': False, 'preserveNewLines': True, 'unformatted': 'wbr', 'extraLiners': 'head, body, /html', 'endWithNewline': False, 'wrapLineLength': 120, 'wrapAttributes': 'auto'}, 'autoClosingTags': True, 'validate': {'scripts': True, 'styles': True}, 'suggest': {'html5': True}}, 'javascript': {'referencesCodeLens': {'showOnAllFunctions': False, 'enabled': False}, 'updateImportsOnFileMove': {'enabled': 'prompt'}, 'implicitProjectConfig': {'experimentalDecorators': False, 'checkJs': True}, 'format': {'insertSpaceAfterOpeningAndBeforeClosingNonemptyBrackets': False, 'semicolons': 'ignore', 'enable': True, 'insertSpaceBeforeAndAfterBinaryOperators': True, 'placeOpenBraceOnNewLineForFunctions': False, 'insertSpaceAfterConstructor': False, 'insertSpaceAfterOpeningAndBeforeClosingTemplateStringBraces': False, 'insertSpaceAfterOpeningAndBeforeClosingNonemptyParenthesis': False, 'insertSpaceBeforeFunctionParenthesis': False, 'insertSpaceAfterFunctionKeywordForAnonymousFunctions': False, 'insertSpaceAfterSemicolonInForStatements': True, 'placeOpenBraceOnNewLineForControlBlocks': False, 'insertSpaceAfterKeywordsInControlFlowStatements': True, 'insertSpaceAfterOpeningAndBeforeClosingJsxExpressionBraces': False, 'insertSpaceAfterOpeningAndBeforeClosingNonemptyBraces': True, 'insertSpaceAfterCommaDelimiter': True}, 'suggestionActions': {'enabled': True}, 'preferences': {'renameShorthandProperties': True, 'importModuleSpecifier': 'auto', 'quoteStyle': 'single'}, 'autoClosingTags': True, 'validate': {'enable': True}, 'suggest': {'autoImports': True, 'completeJSDocs': True, 'enabled': True, 'includeAutomaticOptionalChainCompletions': True, 'paths': True, 'names': True, 'completeFunctionCalls': False}}}}
::  -> lsp-html initialized: {}
lsp-html: Registering request handler for workspace/didChangeWorkspaceFolders failed.
lsp-html: Unhandled exception: client/registerCapability
Error: client/registerCapability
    at handleResponse (C:\Users\Clover\AppData\Local\Sublime Text 3\Cache\LSP-html\vscode-html\node_modules\vscode-jsonrpc\lib\main.js:449:48)
    at processMessageQueue (C:\Users\Clover\AppData\Local\Sublime Text 3\Cache\LSP-html\vscode-html\node_modules\vscode-jsonrpc\lib\main.js:276:17)
    at Immediate.<anonymous> (C:\Users\Clover\AppData\Local\Sublime Text 3\Cache\LSP-html\vscode-html\node_modules\vscode-jsonrpc\lib\main.js:260:13)
    at processImmediate (internal/timers.js:456:21)
```

(we now have formatting disabled at this moment)

**Comment by rchl** [comment]
The issue I'm having is weird:
```
node /Users/me/Library/Application\ Support/Sublime\ Text\ 3/Cache/LSP-html/vscode-html/out/htmlServerMain.js --stdio                                                                1 ↵  2728  09:07:28 
internal/modules/cjs/loader.js:638
    throw err;
    ^

Error: Cannot find module 'typescript'
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:636:15)
    at Function.Module._load (internal/modules/cjs/loader.js:562:25)
    at Module.require (internal/modules/cjs/loader.js:692:17)
    at require (internal/modules/cjs/helpers.js:25:18)
    at Object.<anonymous> (/Users/me/Library/Application Support/Sublime Text 3/Cache/LSP-html/vscode-html/out/modes/javascriptMode.js:10:12)
    at Module._compile (internal/modules/cjs/loader.js:778:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:789:10)
    at Module.load (internal/modules/cjs/loader.js:653:32)
    at tryModuleLoad (internal/modules/cjs/loader.js:593:12)
    at Function.Module._load (internal/modules/cjs/loader.js:585:3)
```
The typescript dependency is in fact not installed in cache directory. If I run `npm i --production` there manually then it works. (I've tried re-saving plugin.py to re-install automatically but that is always missing `typescript` dependency...).

**Comment by jfcherng** [comment]
> The typescript dependency is in fact not installed in cache directory. If I run `npm i --production` there manually then it works. (I've tried re-saving plugin.py to re-install automatically but that is always missing `typescript` dependency...).

Once upon a time, I mixed the `-D/--save/--save-dev` in the build script between LSP-html and LSP-css. Maybe you are suffering from that? If `Cache\LSP-html` is deleted and restart ST, it should be auto fix?

But then the server (in `Cache\LSP-html`) will not be updated even if there is a new release?

**Comment by jfcherng** [comment]
`javascript` errors are not linted. (because there is no `vscode-javascript-languageserver` after all?)

**Comment by rchl** [comment]
> But then will the server (in `Cache/LSP-html`) be updated even if there is a new release?

For when we want to update we have to change `version` in `package.json`:
https://github.com/sublimelsp/lsp_utils/blob/a300b10fefcc0ebb830ca9985963ee7ebbf351a6/st3/lsp_utils/server_npm_resource.py#L76-L80

**Comment by rchl** [comment]
> Once upon a time, I mixed `-D/--save/--save-dev` in the build script between LSP-html and LSP-css. Maybe you are suffering from that? If `Cache/LSP-html` is deleted and restart ST, it should be auto fix?

Right, something like that. It was an issue with me having `vscode_html/node_modules` in the package directory. That whole thing was copied to the cache I guess (but not sure why `npm i` there didn't add it then... Anyways, not important.

File: LSP-html.sublime-settings [hunk 1]
```diff
@@ -13,18 +12,120 @@
...
 	],
 	"initializationOptions": {
-		"provideFormatter": true,
 		"embeddedLanguages": {
 			"css": true,
-			"html": true,
 			"javascript": true,
 		},
+		"dataPaths": [],
+		"provideFormatter": false,
```
Comment (rchl) [inline LSP-html.sublime-settings]: So should it be `true` then?

Comment (jfcherng) [reply LSP-html.sublime-settings]: yeah. i think formatting is a common and useful functionality.

**Pull request merged**